### ISSUE TITLE
Errors when using with vim-vinegar

### ISSUE THREAD

**Issue by lencioni** [kshenoy/vim-signature#104]
I'm using vim-signature along with vim-vinegar and have been getting some errors from vim-signature when browsing around my files using Netrw.

Sometimes when pressing `-` while in Netrw to go up a directory level, I see the following errors reported:

```
E116: Invalid arguments for function stridx(b:SignatureIncludeMarkers, l:char) >= 0
E15: Invalid expression: stridx(b:SignatureIncludeMarkers, l:char) >= 0
```

**Comment by kshenoy** [comment]
That function gets invoked when pressing `m` and grabs the next character that is typed.

Can you figure out what you're pressing when you get the error? That would help me a lot.

**Comment by lencioni** [comment]
I get the error when pressing `-`, but not every single time. On a fresh boot of Vim, if I press `-` it brings up Netrw without any errors. Then, the first time I press `-` to go up a directory, I see the errors, but not on subsequent times.

**Comment by lencioni** [comment]
I'm not sure if this is useful information or not, but I've started seeing some more errors occasionally, when using Netrw, always when pressing `-` to go up a directory level:

```
Error detected while processing function <SNR>63_opendir[7]..<SNR>210_NetrwBrowseUpDir[46]..<SNR>210_NetrwBrowseChgDir[273]..signature#utils#Input:
line    7:
E121: Undefined variable: b:SignatureIncludeMarkers
E116: Invalid arguments for function stridx(b:SignatureIncludeMarkers, l:char) >= 0
E15: Invalid expression: stridx(b:SignatureIncludeMarkers, l:char) >= 0
line   16:
E121: Undefined variable: b:SignatureIncludeMarkers
E116: Invalid arguments for function stridx(b:SignatureIncludeMarkers, l:char) >= 0
E15: Invalid expression: stridx(b:SignatureIncludeMarkers, l:char) >= 0
Press ENTER or type command to continue
```

**Comment by andrewferrier** [comment]
+1. I'm seeing this too.

**Comment by kshenoy** [comment]
I'm sorry you guys are getting this error; I'm still trying to reproduce it at my end without much success.

If you find a definite way to reproduce it please let me know.

**Comment by andrewferrier** [comment]
kshenoy, thanks for the reply. It appears I can now recreate this consistently:

VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Nov  4 2015 17:40:55)
MacOS X (unix) version
Included patches: 1-909

vim-signature commit 7cabfb5
vim-vinegar (https://github.com/tpope/vim-vinegar) commit b8b6de8

Start up vim with a new file: `vim /tmp/xyz.txt`.

Press `-` twice. The first will succeed, and show the `/tmp` directory. The second will fail with:

```
E116: Invalid arguments for function stridx(b:SignatureIncludeMarkers, l:char) >= 0
E15: Invalid expression: stridx(b:SignatureIncludeMarkers, l:char) >= 0
```

Let me know if that's not sufficient, I'll find a more self-contained recreation example.

**Comment by andrewferrier** [comment]
It's not clear to me, by the way, that this is a vim-signature problem as opposed to vim-vinegar :) But hopefully this helps narrow it down a bit though...

**Comment by kshenoy** [comment]
Hello @andrewferrier thanks for providing that, it is definitely a step forward. However, I tried with Vim patched to 516 and it still works for me. I'll try again later with a fully updated version of vim.

**Comment by lencioni** [comment]
If it is helpful, I'm on Vim 7.4 with patches 1-1063.

**Comment by kshenoy** [comment]
Updated to 1246 and still not hitting it. These were the contents of my vimrc:

    filetype plugin indent on
    set rtp+=~/.vim/bundle/vim-signature
    set rtp+=~/.vim/bundle/vim-vinegar

and my .vim setup:

    .vim
    ├─ vimrc
    └─ bundle
       ├─ vim-signature
       └─ vim-vinegar

Can you please try the following setup and see if you still get the error?

**Comment by andrewferrier** [comment]
OK, I just tried the same setup as you (my bundle directory is named with plugged, but otherwise the same scenario) - with the same vim-vinegar and vim-signature commits as above - and I don't see the problem any more. Maybe another plugin interacting... sigh. Let me see if I can narrow it down a bit further...

**Comment by andrewferrier** [comment]
Argh, appears this is more intermittent than I thought, can't reproduce it the way I was before, now. 

Any suggestions for any debug logging or similar I can turn on for when/if it happens again?

**Comment by andrewferrier** [comment]
OK, I've recreated it somewhere else in my directory tree. One thing I have spotted - it only seems to happen when using `-` for the first time in a vim session to reach a particular directory. If I then go into a subdir and back to its parent using `-`, it never shows the error the second time. I'm guessing it's some lazy-loading behavior as netrw loads a particular directory.

**Comment by andrewferrier** [comment]
My vimscript knowledge is a bit basic, but as far as I can tell the error must be coming from the `signature#utils#Input` function in `utils.vim` in vim-signature - that's the only place that uses the line of code referenced in the error. But puzzle - that function seems to only be called when the user is interactively asked for a character. Why would that happen when doing something unrelated (navigating up a tree using vim-vinegar)?

**Comment by andrewferrier** [comment]
OK, I *think* I've found a possible culprit - I think it could be the fact I'm using [vim-plug](https://github.com/junegunn/vim-plug) as my plugin manager.

At least at the moment, I can now consistently reproduce by running `vi ~/.vim/plugged/vim-signature/plugin/signature.vim` and hitting `-` twice (I think the fact it's `signature.vim` I'm editing is mostly incidental), *if* my ~/.vimrc looks like this:

```
call plug#begin()
Plug 'kshenoy/vim-signature'
Plug 'tpope/vim-vinegar'
call plug#end()
```

If, however, I switch it to your example:

```
filetype plugin indent on
set rtp+=~/.vim/plugged/vim-signature
set rtp+=~/.vim/plugged/vim-vinegar
```

I no longer see the problem (`-` works as expected without issue).

By the way, I wondered if it was the load order, so I swapped the first list above inside the vim-plug loading block, but that didn't seem to make a difference.

**Comment by kshenoy** [comment]
@andrewferrier I tried with `vim-plug` as well, using the same vimrc as you. However, I don't get the error.

While I'd like to get to the bottom of this, I understand it might be frustrating for you both. If you want, you can put a hack in to avoid this issue altogether

In plugin/signature.vim, replace [this line](https://github.com/kshenoy/vim-signature/blob/7cabfb5a3d3b45e739eb1d7e198782fb4a5a23da/plugin/signature.vim#L55) with

    autocmd BufEnter,FileType * 
            \ if (&filetype ==? 'nerdtree' || &filetype ==? 'netrw') 
            \|  call signature#utils#Maps('remove') 
            \|endif

Please test this fix. Once I know it works, I'll donate it to trunk.

**Comment by andrewferrier** [comment]
@kshenoy strange that you still cannot reproduce. Are you using a Mac? Wonder if its specific to a certain version of vim.

Anyway, I just tested that patch, and yes, that works - thanks. Actually that doesn't look so hacky to me - I would guess that in practice vim-signature wouldn't even be used in netrw or nerdtree-type buffers anyway - it doesn't really make sense to use signs in such places.

If you want to merge it into trunk, that'd certainly help.

Thanks very much for your attention to this problem - and all your work on the plugin, great job.

**Comment by kshenoy** [comment]
Donated to trunk. Sorry, we couldn't get to the bottom of this.

Closing; reopen if required

**Comment by andrewferrier** [comment]
Perfect, that works for me. Thanks!

**Comment by lencioni** [comment]
Thanks for working on this!

**Comment by lencioni** [comment]
This issue seems to have returned for me recently.

```
Error detected while processing function <SNR>90_NetrwBrowseChgDir[360]..signature#utils#Input:
line    7:
E121: Undefined variable: b:SignatureIncludeMarkers
E15: Invalid expression: (b:SignatureIncludeMarkers =~# l:in)
line   18:
E121: Undefined variable: b:SignatureIncludeMarkers
E15: Invalid expression: (  (b:SignatureIncludeMarkers =~# l:char) && (l:char != ' ') )
```

**Comment by kshenoy** [comment]
Sorry about that. I deleted it because it seemed to be working for me but I forgot that I was unable to reproduce it the last time either. Trying out a different fix this time (#125). Let me know in case of any issues.
